include/nighthawk/user_defined_output/user_defined_output_plugin.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+
+#include "external/envoy/source/common/common/statusor.h"
+#include "external/envoy/source/common/http/header_map_impl.h"
+
+namespace Nighthawk {
+
+// Information about a Nighthawk worker thread. May expand to contain more fields over time as
+// desired.
+struct WorkerMetadata {
+  // Identifies which worker instantiated the plugin instance.
+  int worker_number;
+};
+
+/**
+ * An interface for the UserDefinedOutputPlugin that receives responses and allows users to
+ * attach their own custom output to each worker Result.
+ *
+ * All UserDefinedOutputPlugins must be thread safe, as it may receive multiple responses
+ * simultaneously. In addition, handleResponseData and handleResponseHeaders may be called in any
+ * order or possibly concurrently. GetPerWorkerOutput is guaranteed to be called after
+ * handleResponseData and handleResponseHeaders have been called for every relevant response.
+ *
+ * TODO(dubious90): Throughout file, update comments to contain counter names and other related
+ * specifics.
+ *
+ * TODO(dubious90): Comment on behavior of "relevant responses". e.g. If this plugin still gets
+ * called on pool_overflows or other edge cases
+ */
+class UserDefinedOutputPlugin {
+public:
+  virtual ~UserDefinedOutputPlugin() = default;
+
+  /**
+   * Receives the headers from a single HTTP response, and allows the plugin to collect data based
+   * on those headers.
+   *
+   * Plugins should return statuses for invalid data or when they fail to process the data. Any
+   * non-ok status will be logged and increment a counter that will be added to the worker Result.
+   * Callers can also provide a failure predicate for this counter that will abort the request
+   * after n plugin failures.
+   *
+   * Must be thread safe.
+   *
+   * @param headers
+   */
+  virtual absl::Status
+  handleResponseHeaders(const Envoy::Http::ResponseHeaderMapPtr&& headers) PURE;
```

Comment:
The latest version seems to be passing it by value. We likely don't want to do that - since this is a smart pointer this would mean that Nighthawk would lose the ownership of the object. This should probably become a raw pointer now.

Reply:
Yep, had meant to make it a raw ptr (*) and corrected it incorrectly to Ptr without thinking. Thanks